### PR TITLE
refactor EventSourcedTopology

### PR DESCRIPTION
**Pull request by husseinjoe**
EventSourcedTopology is really big class to be testable.
The idea of this refactoring is to split the logic in this class to 3 pieces

- Process command request and publish resulting events into events' topic.
- Consume generated events, call handlers 
- Update state stores

### CODE REVIEW

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package io.simplesource.kafka.internal.streams.topologies;
+
+import io.simplesource.kafka.api.AggregateResources;
+import io.simplesource.kafka.api.ResourceNamingStrategy;
+
+public class AggregateStreamResourceNames {
```
Comment (szoio) [inline simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java]: Could be a `@Value` class

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    private final ResourceNamingStrategy resourceNamingStrategy;
+    private final String aggregateName;
+
+    public AggregateStreamResourceNames(ResourceNamingStrategy resourceNamingStrategy, String aggregateName) {
```
Comment (szoio) [inline simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java]: If we extract `topicName` and `stateStoreName` methods into a separate class, please can we remove the corresponding methods (and their usages) from the `EventSourcedTopology` class.

Otherwise we're stuck with two ways of doing the same thing.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/EventSourcedTopology.java [hunk 3]
```diff
@@ -1,14 +1,14 @@
 package io.simplesource.kafka.internal.streams;
 
 import io.simplesource.api.Aggregator;
-import io.simplesource.api.InitialValue;
```
Comment (szoio) [inline simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/EventSourcedTopology.java:4]: Maybe we should move this file into the new `topologies` folder.

Also, it feels like we've partially broken this file up. Is this part complete?

There's still a lot going on - I prefer a class / object to be either a worker or a delegator, and preferably not both.

As an alternative to breaking up a big class into smaller ones to make it easier to test, an alternative is to separate the state out completely (into a value class - it's all final), say `TopologyContext`, and then make everything else static. That's my preference, but I realise it's a question of style, what's done here is also valid.

But then I'd like to see this class broken up completely so it only delegates.

**Review by szoio — COMMENTED** [review]
Looks good in general.

Tests for the topology are good and very much needed.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java (hunk 2 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java]: Yes, once I finish with all pieces, all these methods will be removed and only defined in `AggregateStreamResourceNames`

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/EventSourcedTopology.java (hunk 3 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/EventSourcedTopology.java:4]: At the end of this refactoring, `EventSourcedTopology` will be a delegator to all other sub-topologies(`CommandProcessingTopology` is one of them).

This `EventSourcedTopology` has to be public class as it will be accessible by `EventSourcedStreamsApp`. All other sub-topology classes will be package-private

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
+package io.simplesource.kafka.internal.streams.topologies;
```
Comment (szoio) [inline simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java]: I prefer the target type itself to be immutable. `EventSourceTopology` has `addTopology` and `addStateStores` that return `void`, so it kind of looks like a builder itself, but it's not.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventProcessingSubTopology.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                        );
+                    });
+                    return new AggregateUpdateResult<>(
+                            result.commandId(),
+                            result.readSequence(),
+                            aggregateUpdateResult);
+                });
+    }
+
+    private void publishAggregateUpdates(final KStream<K, AggregateUpdateResult<A>> aggregateUpdateStream) {
```
Comment (szoio) [inline simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventProcessingSubTopology.java:54]: If we're going to have an `AggregateUpdateResultStreamConsumer`, should we not use it to consume these?

**Review by szoio — COMMENTED** [review]
I do find the relationship between `StreamsBuilder`, `EventSourcedTopologyBuilder`, `EventSourcedTopology` and `Topology` a little  confusing.

Because of the way StreamsBuilder works, where it itself follows the builder pattern, overlaying further builder pattern implementations on top of it get's a little hard to follow.

What we end up with is effectively:
```
final StreamsBuilder builder = new StreamsBuilder();
new EventSourcedTopologyBuilder<>(aggregateSpec).build().addTopology(builder);
return builder.build();
```
My preference that I would be leaning towards would be to separate out the state from the original `EventSourcingTopology` into some kind of context object (a value class) and then pass that around.
Then everything else is just static functions that we can rearrange how we want, and are easy to test.

Anyway I'm happy to have the test coverage for the topology.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java (hunk 4 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java]: `EventSourcedTopologyBuilder` is only added to create all the dependencies that  `EventSourceTopology` needs.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventProcessingSubTopology.java (hunk 5 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventProcessingSubTopology.java:54]: The way I look to these sub-topology is that it contains source, some processors, and sink nodes. In this case the source is `events` topic(stream passed from command processing sub-topology), and the sink is the publishing to the aggregate update topic.

`AggregateUpdateResultStreamConsumer` are only consumers of a stream.

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java (hunk 1 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/AggregateStreamResourceNames.java]: This got removed and replaced by `AggregateTopologyContext`

File: simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java (hunk 4 above)
Comment (husseinjoe) [reply simplesource-command-kafka/src/main/java/io/simplesource/kafka/internal/streams/topologies/EventSourcedTopologyBuilder.java]: All sub-topology classes are initialised with `AggregateTopologyContext` instead of passing only required dependencies to each sub-topology class. So, no need to have `EventSourcedTopologyBuilder` anymore
